common/flatpak-run.c

```diff
@@ -3230,7 +3232,48 @@ check_parental_controls (const char    *app_ref,
                          _("Running %s is not allowed by the policy set by your administrator"),
                          app_ref);
 
-   return TRUE;
+  /* TODO: is there a better way to get the desktop file/mimetypes here? */
```

Comment:
Maybe using `g_desktop_app_info_new()` and then we could merge the filter checks and use `mct_app_filter_is_appinfo_allowed()` (which already checks by path, content type and appref) instead. But then other parts of the code seem to use this pattern to load the desktop file (nothing using `GAppInfo`/`GDesktopAppInfo`).

Reply:
I don’t know if flatpak will have added the right directory (`${deploy_path}/export/share/applications/`) to the set of paths checked by `g_desktop_app_info_new()` for desktop files by this point, which is probably why other bits of code don’t use `g_desktop_app_info_new()`.

You could use `g_desktop_app_info_new_from_filename()` and pass in the `desktop_file_path` you build below, though, which would save parsing the key file yourself.

Reply:
Makes sense and I will try using `GDesktopAppInfo` here, but still wonder why other parts of the code seem to prefer parsing the key file directly (using `GKeyFile`).

Reply:
Other parts of the existing upstream code, or our downstream modifications? If it’s the existing upstream code, then I guess that’s a question for `#flatpak` on freenode.

Reply:
Checking again, looks like the other places are using `GKeyFile` but to handle other key files (e.g. ostree repo config), not the desktop file. The only place that seems to be opening the desktop file with `GKeyFile` is `tests/testlibrary.c:check_desktop_file`.

Reply:
> Have you looked at how easy it would be to add unit tests for this? I’m wary we’re starting to build up more and more parental controls functionality without any unit tests.

I had a look and to add unit tests we'd need to add a mock accounts service extension for the AppFilter, etc - I'd suggest to handle this as a separate task, what do you think?

Reply:
> I'd suggest to handle this as a separate task, what do you think?

Sure.

Use python-dbusmock for mocking up the accounts-service.

Reply:
Thanks for the review, added https://phabricator.endlessm.com/T27091 for the tests.